src/inmanta/deploy/scheduler.py

```diff
@@ -0,0 +1,163 @@
+    """
+    def __init__(self) -> None:
+        self._state: ModelState = ModelState(version=0)
+        self._work: work.ScheduledWork = work.ScheduledWork(requires=model_state.requires)
+
+        # We uphold two locks to prevent concurrency conflicts between external events (e.g. new version or deploy request)
+        # and the task executor background tasks.
+        #
+        # - lock to block scheduler state access (both model state and scheduled work) during scheduler-wide state updates
+        #   (e.g. trigger deploy). A single lock suffices since all state accesses (both read and write) by the task runners are
+        #   short, synchronous operations (and therefore we wouldn't gain anything by allowing multiple readers).
+        # TODO: this lock may have little value left now. Consider.
+        self._scheduler_lock: asyncio.Lock = asyncio.Lock()
+        # - lock to serialize scheduler state updates (i.e. process new version)
+        self._update_lock: asyncio.Lock = asyncio.Lock()
+
+    def start(self) -> None:
+        # TODO (ticket): read from DB instead
+        pass
+
+    async def deploy(self) -> None:
+        async with self._scheduler_lock:
+            # TODO: more efficient access to dirty set by caching it on the ModelState
+            dirty: Set[ResourceIdStr] = {
+                r for r, details in self._state.resource_state.items() if details.status == ResourceStatus.HAS_UPDATE
+            }
+            # TODO: pass in running deploys
+            self._work.update_state(ensure_scheduled=dirty, running_deploys={})
+
+    # TODO: name
+    # TODO (ticket): design step 2: read new state from DB instead of accepting as parameter (method should be notification only, i.e. 0 parameters)
+    async def new_version(
+        self,
+        version: int,
+        resources: Mapping[ResourceIdStr, ResourceDetails],
+        requires: Mapping[ResourceIdStr, Set[ResourceIdStr]],
+    ) -> None:
+        async with self._update_lock:
+            # Inspect new state and mark resources as "update pending" where appropriate. Since this method is the only writer
+            # for "update pending", and a stale read is acceptable, we can do this part before acquiring the exclusive scheduler
+            # lock.
+            # TODO: what to do when an export changes handler code without changing attributes? Consider in deployed state? What
+            #   does current implementation do?
+            deleted_resources: Set[ResourceIdStr] = self._state.resources.keys() - resources.keys()
+            for resource in deleted_resources:
+                self._work.delete_resource(resource)
+
+            # TODO: make sure this part (before scheduler lock is acquired) doesn't block queue until lock is acquired: either
+            # run on thread or make sure to regularly pass control to IO loop (preferred)
```

Comment:
running it on a thread requires locking and usage of concurrency safe data structures, which can be hard to get any benefit out off (definitely in python)

Reply:
Indeed, I decided against it but hadn't removed the note yet.